crates/services/p2p/src/service.rs

```diff
@@ -857,14 +956,74 @@ impl SharedState {
         let (response_from_peer, response) =
             receiver.await.map_err(|e| anyhow!("{e}"))?;
         assert_eq!(
-            peer_id,
+            peer_id.as_ref(),
             response_from_peer.to_bytes(),
             "Bug: response from non-requested peer"
         );
 
         response.map_err(|e| anyhow!("Invalid response from peer {e:?}"))
     }
 
+    pub async fn get_all_transactions_ids_from_peer(
+        &self,
+        peer_id: FuelPeerId,
+    ) -> anyhow::Result<Vec<TxId>> {
+        let (sender, receiver) = oneshot::channel();
+        let from_peer = PeerId::from_bytes(peer_id.as_ref()).expect("Valid PeerId");
+        let request = TaskRequest::TxPoolGetAllTxIds {
+            from_peer,
+            channel: sender,
+        };
+        self.request_sender.send(request).await?;
+
+        let (response_from_peer, response) =
+            receiver.await.map_err(|e| anyhow!("{e}"))?;
+
+        debug_assert_eq!(
+            peer_id.as_ref(),
+            response_from_peer.to_bytes(),
+            "Bug: response from non-requested peer"
+        );
+
+        let txs = response.map_err(|e| anyhow!("Invalid response from peer {e:?}"))?;
+        if txs.len() > self.max_txs_per_request {
+            return Err(anyhow!("Too many transactions requested: {}", txs.len()));
+        }
+        Ok(txs)
+    }
+
+    pub async fn get_full_transactions_from_peer(
+        &self,
+        peer_id: FuelPeerId,
+        tx_ids: Vec<TxId>,
+    ) -> anyhow::Result<Vec<Option<Transaction>>> {
+        let (sender, receiver) = oneshot::channel();
+        let from_peer = PeerId::from_bytes(peer_id.as_ref()).expect("Valid PeerId");
+        let request = TaskRequest::TxPoolGetFullTransactions {
+            tx_ids,
+            from_peer,
+            channel: sender,
+        };
+        self.request_sender.send(request).await?;
+
+        let (response_from_peer, response) =
+            receiver.await.map_err(|e| anyhow!("{e}"))?;
+        debug_assert_eq!(
+            peer_id.as_ref(),
+            response_from_peer.to_bytes(),
+            "Bug: response from non-requested peer"
+        );
+
+        let txs = response.map_err(|e| anyhow!("Invalid response from peer {e:?}"))?;
+        if txs.len() > self.max_txs_per_request {
+            return Err(anyhow!("Too many transactions requested: {}", txs.len()));
+        }
+        Ok(txs.into_iter().map(|tx| tx.map(|tx_enum| match tx_enum {
+            NetworkableTransactionPool::Transaction(tx) => tx,
+            NetworkableTransactionPool::PoolTransaction(_) => panic!("A transaction that comes from the network must always be deserialized as a Transaction"),
+        })).collect())
```

Comment:
It would be nice to separate iterator part with the `Ok` part.

Also, it would be better to avoid panic, even if it is not possible in the current code. Or implement `From<NetworkableTransactionPool>` to `Transaction`. Another question, is that possible to have one type for serialization and another type for deseriliazation? Maybe we can avoid this check at all

Reply:
I have changed to propagate the result.

I didn't find any way to serialize in a type and deserialize to another with the libp2p enum and serde that's why I ended up with this enum.